src/NUnitTestAdapter/NUnitEngine/NUnitDiscoveryTestClasses.cs

```diff
@@ -0,0 +1,397 @@
+        public virtual bool IsParameterizedMethod => false;
+        public IEnumerable<NUnitProperty> Properties => NUnitDiscoveryProperties.Properties;
+
+        public virtual void AddToAllTestCases(NUnitDiscoveryTestCase tc)
+        {
+            Parent.AddToAllTestCases(tc);
+        }
+    }
+
+
+    public class NUnitDiscoveryTestRun : NUnitDiscoverySuiteBase
+    {
+        public NUnitDiscoveryTestRun(BaseProperties baseProps) : base(baseProps)
+        {
+        }
+
+        public NUnitDiscoveryTestAssembly TestAssembly { get; set; }
+        public override bool IsExplicit =>
+            RunState == RunStateEnum.Explicit || TestAssembly.IsExplicit;
+
+        public void AddTestAssembly(NUnitDiscoveryTestAssembly testAssembly)
+        {
+            TestAssembly = testAssembly;
+        }
+    }
+
+    public class NUnitDiscoveryProperties
+    {
+        private List<NUnitProperty> TheProperties { get; } = new List<NUnitProperty>();
+        public IEnumerable<NUnitProperty> Properties => TheProperties;
+
+        public void Add(NUnitProperty p) => TheProperties.Add(p);
+
+        public bool AllInternal => TheProperties.All(o => o.IsInternal);
+    }
+
+    public sealed class NUnitDiscoveryTestSuite : NUnitDiscoveryCanHaveTestFixture
+    {
+        private readonly List<NUnitDiscoveryTestSuite> testSuites = new List<NUnitDiscoveryTestSuite>();
+
+        private readonly List<NUnitDiscoveryGenericFixture> genericFixtures = new List<NUnitDiscoveryGenericFixture>();
+        private readonly List<NUnitDiscoverySetUpFixture> setUpFixtures = new List<NUnitDiscoverySetUpFixture>();
+        private readonly List<NUnitDiscoveryParameterizedTestFixture> parameterizedFixtures = new List<NUnitDiscoveryParameterizedTestFixture>();
+        public NUnitDiscoveryTestSuite(BaseProperties theBase, INUnitDiscoverySuiteBase parent) : base(theBase, parent)
+        {
+        }
+
+        public IEnumerable<NUnitDiscoveryTestSuite> TestSuites => testSuites;
+        public IEnumerable<NUnitDiscoverySetUpFixture> SetUpFixtures => setUpFixtures;
+        public IEnumerable<NUnitDiscoveryParameterizedTestFixture> ParameterizedFixtures => parameterizedFixtures;
+
+        public IEnumerable<NUnitDiscoveryGenericFixture> GenericFixtures => genericFixtures;
+
+        public override bool IsExplicit
+        {
+            get
+            {
+                var fullList = new List<NUnitDiscoverySuiteBase>();
+                fullList.AddRange(TestFixtures.Cast<NUnitDiscoverySuiteBase>());
+                fullList.AddRange(TestSuites.Cast<NUnitDiscoverySuiteBase>());
+                fullList.AddRange(GenericFixtures.Cast<NUnitDiscoverySuiteBase>());
+                fullList.AddRange(SetUpFixtures.Cast<NUnitDiscoverySuiteBase>());
+                fullList.AddRange(ParameterizedFixtures.Cast<NUnitDiscoverySuiteBase>());
+                return RunState == RunStateEnum.Explicit || fullList.All(o => o.IsExplicit);
```

Comment:
For performance, consider using .Concat instead of buffering into a list.

```cs
return RunState == RunStateEnum.Explicit || TestFixtures.Concat<NUnitDiscoverySuiteBase>(TestSuites)
    .Concat(GenericFixtures)
    .Concat(SetUpFixtures)
    .Concat(ParameterizedFixtures)
    .All(o => o.IsExplicit);
```

Or


```cs
return RunState == RunStateEnum.Explicit || new IEnumerable<NUnitDiscoverySuiteBase>[]
    {
        TestFixtures, TestSuites, GenericFixtures, SetUpFixtures, ParameterizedFixtures
    }.SelectMany(tests => tests).All(o => o.IsExplicit);
```

Reply:
Sure, although I could not see any performance issues around this.  I did check ;-)  But your first suggestion here is better wrt to readability of the intent.

Reply:
The concat struggles with these lists.......

Reply:
Oh, I didn't realize it was newer than 3.5. Sorry!